streamer/src/bidirectional_channel.rs

```diff
@@ -0,0 +1,396 @@
+        // check if socket is registered;
+        let id = {
+            let data = self.data.read();
+            if let Ok(data) = data {
+                data.sender_socket_address_map
+                    .get(quic_address)
+                    .map_or(0, |x| *x)
+            } else {
+                0
+            }
+        };
+        if id == 0 {
+            return;
+        }
+        let data = self.data.write();
+        if let Ok(mut data) = data {
+            let data = &mut *data;
+            packets.iter().for_each(|packet| {
+                let signature = get_signature_from_packet(packet);
+                signature.map(|x| data.transaction_signature_map.insert(x, id));
+            });
+        }
+    }
+
+    // this method will start bidirectional relay service
+    // the the message sent to bidirectional service,
+    // will be dispactched to the appropriate sender channel
+    // depending on transcation signature or message hash
+    pub fn serve(&self) {
+        let service_reciever = self.service_reciever.clone();
+        let subscription_data = self.data.clone();
+        tokio::spawn(async move {
+            loop {
+                let service_reciever = service_reciever.clone();
+                let bidirectional_message = service_reciever.recv();
+                match bidirectional_message {
+                    Ok(bidirectional_message) => {
+                        let data = subscription_data.read();
+
+                        if let Ok(data) = data {
+                            let send_stream = {
+                                // if the message has transaction signature then find stream from transaction signature
+                                // else find stream by packet hash
+                                let send_stream_id = data
+                                    .transaction_signature_map
+                                    .get(&bidirectional_message.transaction_signature)
+                                    .map(|x| *x);
+                                send_stream_id.and_then(|x| data.sender_ids_map.get(&x))
+                            };
+                            if let Some(send_stream) = send_stream {
+                                match send_stream.send(bidirectional_message) {
+                                    Err(e) => {
+                                        warn!(
+                                            "Error sending a bidirectional message {}",
+                                            e.to_string()
+                                        );
+                                    }
+                                    _ => {}
+                                }
+                            }
+                        }
+                    }
+                    Err(e) => {
+                        // recv error
+                        warn!("got {} on quic bidirectional channel", e.to_string());
```

Comment:
better use metrics, not sure how often this error can be triggered in production